File: 3042/go/main.go
package main

import "fmt"

type Node struct {
	children map[rune]*Node
	isEnd    bool
}

func NewNode() *Node {
	return &Node{children: make(map[rune]*Node)}
}

type Trie struct {
	root *Node
}

func initTrie() *Trie {
	return &Trie{root: NewNode()}
}

func countPrefixSuffixPairs(words []string) int {

	count := 0

	for i, word := range words {
		for j := i + 1; j < len(words); j++ {
			sub := words[j]

			if isPrefixAndSuffix(word, sub) {
				count++
			}
		}
	}

	return count
}

func main() {
	fmt.Println(countPrefixSuffixPairs([]string{"a", "aba", "ababa", "aa"}))
	fmt.Println(countPrefixSuffixPairs([]string{"pa", "papa", "ma", "mama"}))
	fmt.Println(countPrefixSuffixPairs([]string{"ababa", "ab"}))
}

File: 3042/python/main.py
from typing import *


class Solution:
    def isPrefixAndSuffix(self, substring: str, string: str) -> bool:
        sublen = len(substring)
        return string[:sublen] == substring and string[-sublen:] == substring

    def countPrefixSuffixPairs(self, words: List[str]) -> int:
        count = 0

        for i, word in enumerate(words):
            for j in range(i + 1, len(words)):
                if self.isPrefixAndSuffix(word, words[j]):
                    count += 1
        return count


s = Solution()
print(s.countPrefixSuffixPairs(["a", "aba", "ababa", "aa"]))
print(s.countPrefixSuffixPairs(["pa", "papa", "ma", "mama"]))
print(s.countPrefixSuffixPairs(["abab", "ab"]))
